gr-digital/lib/decision_feedback_equalizer_impl.cc

```diff
@@ -0,0 +1,267 @@
+                                             std::vector<int>{ sizeof(gr_complex),
+                                                               (int)((num_taps_forward +
+                                                                      num_taps_feedback) *
+                                                                     sizeof(gr_complex)),
+                                                               sizeof(unsigned short) }),
+                         sps),
+      filter::kernel::fir_filter_ccc(
+          sps,
+          vector<gr_complex>(num_taps_forward + num_taps_feedback, gr_complex(0, 0))),
+      d_num_taps_fwd(num_taps_forward),
+      d_num_taps_rev(num_taps_feedback),
+      d_sps(sps),
+      d_alg(alg),
+      d_adapt_after_training(adapt_after_training),
+      d_training_sequence(training_sequence),
+      d_training_start_tag(training_start_tag),
+      d_num_taps(num_taps_forward + num_taps_feedback),
+      d_decision_history(num_taps_feedback),
+      d_new_taps(num_taps_forward + num_taps_feedback),
+      d_updated(false),
+      d_training_sample(0),
+      d_filt_buf(num_taps_forward + num_taps_feedback)
+{
+    // not reversed at this point
+    d_new_taps[0] = gr_complex(1.0, 0.0); // help things along with initialized taps
+
+    if (training_start_tag == "" || training_sequence.empty()) {
+        d_training_state = equalizer_state_t::DD;
+    } else {
+        d_training_state = equalizer_state_t::IDLE;
+    }
+
+    set_history(d_num_taps_fwd);
+
+    alg->initialize_taps(d_new_taps);
+    // NOTE: the filter kernel reverses the taps internally
+    filter::kernel::fir_filter_ccc::set_taps(d_new_taps);
+}
+
+void decision_feedback_equalizer_impl::update_decision_history(gr_complex decision)
+{
+    std::rotate(d_decision_history.begin(),
+                d_decision_history.begin() + 1,
+                d_decision_history.end());
+    d_decision_history[d_decision_history.size() - 1] = decision;
+}
+
+int decision_feedback_equalizer_impl::work(int noutput_items,
+                                           gr_vector_const_void_star& input_items,
+                                           gr_vector_void_star& output_items)
+{
+
+    auto in = static_cast<const gr_complex*>(input_items[0]);
+    auto out = static_cast<gr_complex*>(output_items[0]);
+
+    int outlen = output_items.size();
+
+    unsigned short* state = nullptr;
+    gr_complex* taps = nullptr;
+
+    if (outlen > 1)
+        taps = static_cast<gr_complex*>(output_items[1]);
+    if (outlen > 2)
+        state = static_cast<unsigned short*>(output_items[2]);
+
+    if (d_updated) {
+        gr::thread::scoped_lock guard(d_mutex);
+        d_taps = d_new_taps;
+        set_history(d_taps.size());
+        d_updated = false;
+        return 0; // history requirements may have changed.
+    }
+
+    unsigned long int nread = nitems_read(0);
+    vector<tag_t> tags;
+    vector<unsigned int> training_start_samples;
```

Comment:
Probably want to do `training_start_samples.reserve(tags.size())` before looping over the tags. If you think it may affect performance in a good way (because zeroing the values it may make it worse) maybe go so far as to create it at the correct size to start with, and set the values by index instead of `push_back`.